### PR TITLE
Random number generation seeding structure refactor

### PR DESCRIPTION
**Pull request by rnugent3**

### CODE REVIEW

**Comment by rnugent3** [comment]
Checking build and tests on GitHub because ACE-IT told me no. Please hold off review.

File: HEC.FDA.Model/metrics/SystemPerformanceResults.cs [hunk 1]
```diff
@@ -18,10 +18,8 @@ public class SystemPerformanceResults : ValidationErrorLogger
         #region Fields
         private const string AEP_ASSURANCE_TYPE = "AEP";
         private const string STAGE_ASSURANCE_TYPE = "STAGE";
-        private const string AEP_ASSURANCE_FOR_PLOTTING = "AEP_PLOT";
```
Comment (Brennan1994) [inline HEC.FDA.Model/metrics/SystemPerformanceResults.cs:21]: This seems impactful for plotting. We were plotting using a different histogram for some reason. Maybe that's why our plot was all stair steppy

File: HEC.FDA.Model/paireddata/IPairedDataProducer.cs [hunk 2]
```diff
@@ -2,6 +2,7 @@ namespace HEC.FDA.Model.paireddata
 {
     public interface IPairedDataProducer
     {
-        PairedData SamplePairedData(double probability, bool computeIsDeterministic);
+        PairedData SamplePairedData(long iteration, bool computeIsDeterministic);
```
Comment (Brennan1994) [inline HEC.FDA.Model/paireddata/IPairedDataProducer.cs:5]: Do we need to keep this compute is deterministic?

File: HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs [hunk 3]
```diff
@@ -343,18 +346,13 @@ private PairedData IdentifyCentralStageFrequencyAtIndexLocation()
         /// <param name="inventoryAndWaterTupled"></param>
         /// <param name="profileProbabilities"></param>
         /// <returns></returns>
-        private List<StudyAreaConsequencesBinned> ComputeDamageWithUncertaintyAllCoordinates(string damageCategory, IProvideRandomNumbers randomProvider, (Inventory, List<float[]>) inventoryAndWaterTupled, List<double> profileProbabilities)
+        private List<StudyAreaConsequencesBinned> ComputeDamageWithUncertaintyAllCoordinates(string damageCategory, (Inventory, List<float[]>) inventoryAndWaterTupled, List<double> profileProbabilities, bool computeIsDeterministic)
```
Comment (Brennan1994) [inline HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs:349]: Want to talk about whether we still need this bool.

File: HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs [hunk 4]
```diff
@@ -637,7 +635,7 @@ public void ReportProgress(object sender, ProgressReportEventArgs e)
         internal List<string> ProduceImpactAreaStructureDetails(Dictionary<int, string> impactAreaNames)
         {
             //this list will be the size of the number of structures + 1 where the first string is the header
-            List<DeterministicOccupancyType> deterministicOccupancyTypes = Inventory.SampleOccupancyTypes(new compute.MedianRandomProvider(), computeIsDeterministic: true);
+            List<DeterministicOccupancyType> deterministicOccupancyTypes = Inventory.SampleOccupancyTypes(iteration:1, computeIsDeterministic: true);
```
Comment (Brennan1994) [inline HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs:638]: Looks like we previously wanted the mean here. now we're getting the iteration 1 sample. Are we still good?

File: HEC.FDA.Model/compute/ImpactAreaScenarioSimulation.cs [hunk 5]
```diff
@@ -124,18 +127,62 @@ public ImpactAreaScenarioResults Compute(IProvideRandomNumbers randomProvider, C
...
             ReportMessage(this, beginComputeMessageArgs);
-            ComputeIterations(convergenceCriteria, randomProvider, masterseed, computeWithDamage, computeIsDeterministic, cancellationToken);
+            ComputeIterations(convergenceCriteria, computeWithDamage, computeIsDeterministic, cancellationToken);
             _ImpactAreaScenarioResults.ParallelResultsAreConverged(.95, .05);
             MessageEventArgs endComputeMessageArgs = new(new Message($"EAD and performance compute for the impact area with ID {_ImpactAreaID} has completed successfully" + Environment.NewLine));
             ReportMessage(this, endComputeMessageArgs);
             return _ImpactAreaScenarioResults;
         }
 
+        private void PopulateRandomNumbers(ConvergenceCriteria convergenceCriteria)
```
Comment (Brennan1994) [inline HEC.FDA.Model/compute/ImpactAreaScenarioSimulation.cs:137]: I'm interested in whether it makes sense to have each individual object track their seeds, or for ImpactAreaScenarioSimulation to track them for them and inject the right one.  What were your thoughts on that?

**Review by Brennan1994 — CHANGES_REQUESTED** [review]

**Comment by Brennan1994** [comment]
There's more here than i can chew on. Lets chat about a few things in your office today, and we'll approve or identify changes after that

File: HEC.FDA.Model/metrics/SystemPerformanceResults.cs (hunk 1 above)
Comment (rnugent3) [reply HEC.FDA.Model/metrics/SystemPerformanceResults.cs:21]: resolved in person

File: HEC.FDA.Model/paireddata/IPairedDataProducer.cs (hunk 2 above)
Comment (rnugent3) [reply HEC.FDA.Model/paireddata/IPairedDataProducer.cs:5]: resolved

File: HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs (hunk 4 above)
Comment (rnugent3) [reply HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs:638]: Add comments to note that iteration is ignored for compute is deterministic

File: HEC.FDA.Model/compute/ImpactAreaScenarioSimulation.cs (hunk 5 above)
Comment (rnugent3) [reply HEC.FDA.Model/compute/ImpactAreaScenarioSimulation.cs:137]: resolved

File: HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs (hunk 4 above)
Comment (rnugent3) [reply HEC.FDA.Model/stageDamage/ImpactAreaStageDamage.cs:638]: I have added public XML comments on all of the objects that can be sampled

**Review by Brennan1994 — APPROVED** [review]

**Comment by Brennan1994** [comment]
The suspected result changing modification here is moving from a sampling schema which samples a small number of samples from a large number of random number generators to a new schema of a single random number generator with many samples. 

The results appear better distributed after the changes in this request. The design choice between these two sampling schemes was presented in a code review with computer scientists outside the HEC-FDA team, and checked against industry best practices as documented by Microsoft. Generating many values from a single generator is the proper use of _Random.cs_. 

"Both to improve performance and to avoid inadvertently creating separate random number generators that generate identical numeric sequences, we recommend that you create one Random object to generate many random numbers over time, instead of creating new Random objects to generate one random number." - Microsoft docs
https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-random

**Pull request merged**